chemprop/models/ffn.py

```diff
@@ -0,0 +1,439 @@
+from typing import List, Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import re
+
+
+class MultiReadout(nn.Module):
+    """A :class:`MultiReadout` contains a list of FFN for each atom/bond targets prediction."""
+
+    def __init__(
+        self,
+        atom_features_size: int,
+        bond_features_size: int,
+        atom_hidden_size: int,
+        bond_hidden_size: int,
+        num_layers: int,
+        output_size: int,
+        dropout: float,
+        activation: nn.Module,
+        atom_constraints: List[bool] = None,
+        bond_constraints: List[bool] = None,
+        shared_ffn: bool = True,
+        weights_ffn_num_layers: int = 2,
+    ):
+        """
+        :param atom_features_size: Dimensionality of input atomic features.
+        :param bond_features_size: Dimensionality of input bond features.
+        :param atom_hidden_size: Dimensionality of atomic hidden layers.
+        :param bond_hidden_size: Dimensionality of bond hidden layers.
+        :param num_layers: Number of layers in FFN.
+        :param output_size: The size of output.
+        :param dropout: Dropout probability.
+        :param activation: Activation function.
+        :param atom_constraints: A list of booleans indicating whether constraints applied to output of atomic properties.
+        :param bond_constraints: A list of booleans indicating whether constraints applied to output of bond properties.
+        :param shared_ffn: Whether to share weights in the ffn between different atom tasks and bond tasks.
+        :param weights_ffn_num_layers: Number of layers in FFN for determining weights used to correct the constrained targets.
+        """
+        super().__init__()
+        self.shared_ffn = shared_ffn
+
+        if num_layers > 1 and shared_ffn:
+            self.atom_ffn_base = nn.Sequential(
+                build_ffn(
+                    first_linear_dim=atom_features_size,
+                    hidden_size=atom_hidden_size,
+                    num_layers=num_layers - 1,
+                    output_size=atom_hidden_size,
+                    dropout=dropout,
+                    activation=activation,
+                ),
+                activation,
+            )
+            self.bond_ffn_base = nn.Sequential(
+                build_ffn(
+                    first_linear_dim=2*bond_features_size,
+                    hidden_size=bond_hidden_size,
+                    num_layers=num_layers - 1,
+                    output_size=bond_hidden_size,
+                    dropout=dropout,
+                    activation=activation,
+                ),
+                activation,
+            )
+        else:
+            self.atom_ffn_base = nn.Identity()
+            self.bond_ffn_base = nn.Identity()
```

Comment:
why assign these to the identity function if they won't be used? if `shared_ffn` is `False`, it's passed to all initializers and then these objects are ignored

Reply:
Yes... I will replace those with `None`